### PR TITLE
Fix ConcurrentModificationException during Cache eviction

### LINKED ISSUE #89: cache concurrent modification exception
I've noticed this while trying to repro issue #86 .

```
2020-02-03 21:07:37.721 11904-11904/com.dropbox.stor4.bug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.dropbox.stor4.bug, PID: 11904
    java.util.ConcurrentModificationException
        at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:757)
        at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:780)
        at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:185)
        at com.dropbox.android.external.cache4.RealCache.evictEntries(RealCache.kt:218)
        at com.dropbox.android.external.cache4.RealCache.put(RealCache.kt:147)
        at com.dropbox.android.external.store4.impl.RealStore$stream$2.invokeSuspend(RealStore.kt:114)
        at com.dropbox.android.external.store4.impl.RealStore$stream$2.invoke(Unknown Source:10)
        at kotlinx.coroutines.flow.FlowKt__TransformKt$onEach$$inlined$unsafeTransform$1$2.emit(Collect.kt:138)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:33)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:33)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:33)
        at kotlinx.coroutines.flow.FlowKt__ErrorsKt$catchImpl$$inlined$collect$1.emit(Collect.kt:138)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:33)
        at com.dropbox.flow.multicast.Multicaster$newDownstream$2$invokeSuspend$$inlined$transform$1$1.emit(Collect.kt:138)
        at kotlinx.coroutines.flow.FlowKt__ChannelsKt.emitAll(Channels.kt:56)
        at kotlinx.coroutines.flow.FlowKt.emitAll(Unknown Source:1)
        at kotlinx.coroutines.flow.FlowKt__ChannelsKt$emitAll$1.invokeSuspend(Unknown Source:10)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:561)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:727)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:667)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:655)
```

https://github.com/atonamy/store4-issue

To reproduce, checkout the app and hit refresh couple of times and eventually happens. Sorry i know it is not great repro but i'm focusing on the other bug right now so wanted to report this w/o being side-tracked.
.

### PR DESCRIPTION
**Pull request by ychescale9**
Fixes #89.

PR fixes a `ConcurrentModificationException` during cache eviction due to iterating over the `accessQueue` when calling `accessQueue.first()` without synchronizing the `accessQueue` (MutableSet).

Manually verified that the exception is no longer thrown after pressing "REPEAT TEST" multiple times in the [sample app](https://github.com/atonamy/store4-issue).

### CODE REVIEW

File: cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt [hunk 1]
```diff
@@ -215,7 +215,9 @@ internal class RealCache<Key : Any, Value : Any>(
         cleanUpDeadEntries()
 
         while (cacheEntries.size > maxSize) {
-            val entryToEvict = accessQueue.first()
+            val entryToEvict = synchronized(accessQueue) {
```
Comment (yigit) [inline cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt]: sorry i think i'm being stupid here (tired, end of day :/) but where is the iteration over the list?

also, can we at least add a stress test to this code. something that starts w/ a randomly selected seed generated numers and than adds/deletes them in parallel (see might help w/ repro but probably wont :/ ) maybe someone has better ideas on how to test this and possible other issues similar to it.

Comment (ychescale9) [reply cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt]: Sorry it's not obvious without IDE. `first()` is a kotlin extension on `Iterable` which uses iterator internally.

Yeah I'm not sure how to write such test that will consistently fail with `ConcurrentModificationException`. I'll see if I can come up with something.

Comment (yigit) [reply cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt]: but where is the outer iterator? where does it get modified, some other function ? if so, don't they need to sync over it as well? i see there are some multiple locks in the code so it seems hard to ensuure that we are in full setup and also won't deadlock. should we consider consalidating these locks into 1? would that be a perf problem?

Comment (ychescale9) [reply cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt]: `_Collections.kt` from Kotlin stdlib:

```kotlin
/**
 * Returns first element.
 * @throws [NoSuchElementException] if the collection is empty.
 */
public fun <T> Iterable<T>.first(): T {
    when (this) {
        is List -> return this.first()
        else -> {
            val iterator = iterator()
            if (!iterator.hasNext())
                throw NoSuchElementException("Collection is empty.")
            return iterator.next()
        }
    }
}
```

Is this what you're asking?

We have 3 collections in this class:

1. `cacheEntries: MutableMap<Key, CacheEntry<Key, Value>>` which is a concurrent hash map so we are not doing any syncing on it.
2. `writeQueue: MutableSet<CacheEntry<Key, Value>>` which is synchronized using [`Collections.synchronizedSet()`](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedSet(java.util.Set)), and therefore requires explicit synchronization when iterating it. It's not ideal that we have to know the implementation / initialization details of the `MutableSet` to know that it needs to be synchronized before iterating, but I couldn't find a thread-safe `ConcurrentHashSet` (or similar) out of box.
2. `accessQueue: MutableSet<CacheEntry<Key, Value>>` - same as above

The other lock is for key-based synchronization when running cache loader.

Comment (ychescale9) [reply cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt]: I managed to reproduce the `ConcurrentModificationException` by launching multiple coroutines concurrently and writing new cache entries. Added test [1292bd6](https://github.com/dropbox/Store/pull/90/commits/1292bd6741fc1bd42ff373fbbebe6ce4866b719c).

**Pull request merged**